File: Heap/Java/Kth Largest Sum in a Binary Tree.java
// Problem link: https://leetcode.com/problems/kth-largest-sum-in-a-binary-tree/description/
// Time: O(n * log h)
// Space: O(n)
class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        PriorityQueue<Long> minHeap = new PriorityQueue<>();
        Queue<TreeNode> queue = new LinkedList<>();
        long total;
        TreeNode node;
        int size;

        queue.offer(root);

        while(!queue.isEmpty()) {

            total = 0;
            size = queue.size();

            for(int itr = 0; itr < size; itr++) {

                node = queue.poll();
                total += node.val;

                if(node.left != null) {
                    queue.offer(node.left);
                }

                if(node.right != null) {
                    queue.offer(node.right);
                }
            }

            minHeap.offer(total);

            if(minHeap.size() > k) {
                minHeap.poll();
            }
        }

        return minHeap.size() == k ? minHeap.peek() : -1;
    }
}

File: Heap/Python/Kth Largest Sum in a Binary Tree.py
# Problem link: https://leetcode.com/problems/kth-largest-sum-in-a-binary-tree/description/
# Time: O(n * log h)
# Space: O(n)
class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        minHeap = []

        queue = deque([root])

        while queue:

            total = 0

            for _ in range(len(queue)):

                node = queue.popleft()
                total += node.val

                if node.left:
                    queue.append(node.left)
                
                if node.right:
                    queue.append(node.right)

            heappush(minHeap, total)

            if len(minHeap) > k:
                heappop(minHeap)

        return minHeap[0] if len(minHeap) == k else -1
